# DP: Add multiarch support to GCC.
# DP:
# DP: Convert the multilib option to a target triplet,
# DP: add multiarch include directories and libraries path:
# DP:	/usr/local/include/<arch>-linux-gnu
# DP:	/usr/include/<arch>-linux-gnu
# DP:	/usr/lib/<arch>-linux-gnu
# DP: to the system paths.

2012-11-14  Matthias Klose  <doko@ubuntu.com>

	* config/alpha/t-linux: New file; define MULTIARCH_DIRNAME.
	* config.gcc <alpha*-*-linux*> (tmake_file): Include alpha/t-linux.

2012-11-14  Matthias Klose  <doko@ubuntu.com>

	* config/arm/t-linux-eabi: Define MULTIARCH_DIRNAME for linux target.

2012-11-14  Matthias Klose  <doko@ubuntu.com>

	* config/ia64/t-linux: New file; define MULTIARCH_DIRNAME.
	* config.gcc <ia64*-*-linux*> (tmake_file): Include ia64/t-linux.

2012-11-14  Matthias Klose  <doko@ubuntu.com>

	* config/m68k/t-linux: Define MULTIARCH_DIRNAME.

2012-11-14  Matthias Klose  <doko@ubuntu.com>

	* config/mips/t-linux64: Add multiarch names in MULTILIB_OSDIRNAMES.

2012-11-14  Matthias Klose  <doko@ubuntu.com>

	* config/rs6000/t-linux64: Add multiarch names in MULTILIB_OSDIRNAMES.
	* config/rs6000/t-linux: New file; define MULTIARCH_DIRNAME.
	* config/rs6000/t-spe: Define MULTIARCH_DIRNAME.
	* config.gcc <powerpc-*-linux* | powerpc64-*-linux*> (tmake_file):
	Include rs6000/t-linux for 32bit non-biarch configurations.

2012-11-14  Matthias Klose  <doko@ubuntu.com>

	* config/s390/t-linux64: Add multiarch names in MULTILIB_OSDIRNAMES.

--- a/src/gcc/config/mips/t-linux64	(revision 193508)
+++ b/src/gcc/config/mips/t-linux64	(working copy)
@@ -18,4 +18,9 @@
 
 MULTILIB_OPTIONS = mabi=n32/mabi=32/mabi=64
 MULTILIB_DIRNAMES = n32 32 64
-MULTILIB_OSDIRNAMES = ../lib32 ../lib ../lib64
+MIPS_EL = $(if $(filter %el, $(firstword $(subst -, ,$(target)))),el)
+MIPS_SOFT = $(if $(strip $(filter MASK_SOFT_FLOAT_ABI, $(target_cpu_default)) $(filter soft, $(with_float))),soft)
+MULTILIB_OSDIRNAMES = \
+	../lib32$(call if_multiarch,:mips64$(MIPS_EL)-linux-gnuabin32$(MIPS_SOFT)) \
+	../lib$(call if_multiarch,:mips$(MIPS_EL)-linux-gnu$(MIPS_SOFT)) \
+	../lib64$(call if_multiarch,:mips64$(MIPS_EL)-linux-gnuabi64$(MIPS_SOFT))
--- a/src/gcc/config/arm/t-linux-eabi	(revision 193508)
+++ b/src/gcc/config/arm/t-linux-eabi	(working copy)
@@ -24,3 +24,6 @@
 #MULTILIB_OPTIONS     += mcpu=fa606te/mcpu=fa626te/mcpu=fmp626/mcpu=fa726te
 #MULTILIB_DIRNAMES    += fa606te fa626te fmp626 fa726te
 #MULTILIB_EXCEPTIONS  += *mthumb/*mcpu=fa606te *mthumb/*mcpu=fa626te *mthumb/*mcpu=fmp626 *mthumb/*mcpu=fa726te*
+
+ARM_EB = $(if $(findstring TARGET_BIG_ENDIAN_DEFAULT=1, $(tm_defines)),eb)
+MULTIARCH_DIRNAME = $(call if_multiarch,arm$(ARM_EB)-linux-gnueabi$(if $(filter hard,$(with_float)),hf))
--- a/src/gcc/config/alpha/t-linux	(revision 0)
+++ b/src/gcc/config/alpha/t-linux	(revision 0)
@@ -0,0 +1 @@
+MULTIARCH_DIRNAME = $(call if_multiarch,alpha-linux-gnu)
--- a/src/gcc/config/s390/t-linux64	(revision 193508)
+++ b/src/gcc/config/s390/t-linux64	(working copy)
@@ -7,4 +7,5 @@
 
 MULTILIB_OPTIONS = m64/m31
 MULTILIB_DIRNAMES = 64 32
-MULTILIB_OSDIRNAMES = ../lib64 $(if $(wildcard $(shell echo $(SYSTEM_HEADER_DIR))/../../usr/lib32),../lib32,../lib)
+MULTILIB_OSDIRNAMES = ../lib64$(call if_multiarch,:s390x-linux-gnu)
+MULTILIB_OSDIRNAMES += $(if $(wildcard $(shell echo $(SYSTEM_HEADER_DIR))/../../usr/lib32),../lib32,../lib)$(call if_multiarch,:s390-linux-gnu)
--- a/src/gcc/config/rs6000/t-linux64	(revision 193508)
+++ b/src/gcc/config/rs6000/t-linux64	(working copy)
@@ -29,4 +29,5 @@
 MULTILIB_OPTIONS        = m64/m32
 MULTILIB_DIRNAMES       = 64 32
 MULTILIB_EXTRA_OPTS     = fPIC
-MULTILIB_OSDIRNAMES	= ../lib64 $(if $(wildcard $(shell echo $(SYSTEM_HEADER_DIR))/../../usr/lib32),../lib32,../lib)
+MULTILIB_OSDIRNAMES	= ../lib64$(call if_multiarch,:powerpc64-linux-gnu)
+MULTILIB_OSDIRNAMES    += $(if $(wildcard $(shell echo $(SYSTEM_HEADER_DIR))/../../usr/lib32),../lib32,../lib)$(call if_multiarch,:powerpc-linux-gnu)
--- a/src/gcc/config/rs6000/t-spe	(revision 193508)
+++ b/src/gcc/config/rs6000/t-spe	(working copy)
@@ -71,3 +71,6 @@
 			  mabi=altivec/mlittle \
 			  maltivec/mlittle \
 			  maltivec/mabi=altivec/mlittle
+
+MULTIARCH_DIRNAME = powerpc-linux-gnuspe$(if $(findstring rs6000/e500-double.h, $(tm_file)),,v1)
+
--- a/src/gcc/config/rs6000/t-linux	(revision 0)
+++ b/src/gcc/config/rs6000/t-linux	(revision 0)
@@ -0,0 +1 @@
+MULTIARCH_DIRNAME = powerpc-linux-gnu
--- a/src/gcc/config/sh/t-linux	(revision 193508)
+++ b/src/gcc/config/sh/t-linux	(working copy)
@@ -1,2 +1,4 @@
 MULTILIB_DIRNAMES= 
 MULTILIB_MATCHES = 
+
+MULTILIB_OSDIRNAMES = sh4-linux-gnu:sh4-linux-gnu sh4_nofpu-linux-gnu:sh4-linux-gnu
--- a/src/gcc/config/ia64/t-linux	(revision 0)
+++ b/src/gcc/config/ia64/t-linux	(revision 0)
@@ -0,0 +1 @@
+MULTIARCH_DIRNAME = $(call if_multiarch,ia64-linux-gnu)
--- a/src/gcc/config/m68k/t-linux	(revision 193508)
+++ b/src/gcc/config/m68k/t-linux	(working copy)
@@ -19,6 +19,8 @@
 # Only include multilibs for 680x0 and ColdFire CPUs with an MMU.
 M68K_MLIB_CPU += && ((CPU ~ "^m680") || (CPU ~ "^mcf")) && (FLAGS ~ "FL_MMU")
 
+MULTIARCH_DIRNAME = $(call if_multiarch,m68k-linux-gnu)
+
 # This rule uses MULTILIB_MATCHES to generate a definition of
 # SYSROOT_SUFFIX_SPEC.
 sysroot-suffix.h: $(srcdir)/config/m68k/print-sysroot-suffix.sh
--- a/src/gcc/config.gcc	(revision 193508)
+++ b/src/gcc/config.gcc	(working copy)
@@ -831,6 +831,7 @@
 	;;
 alpha*-*-linux*)
 	tm_file="elfos.h ${tm_file} alpha/elf.h alpha/linux.h alpha/linux-elf.h glibc-stdint.h"
+	tmake_file="${tmake_file} alpha/t-linux"
 	extra_options="${extra_options} alpha/elf.opt"
 	;;
 alpha*-*-freebsd*)
@@ -1550,7 +1553,7 @@
 	;;
 ia64*-*-linux*)
 	tm_file="${tm_file} dbxelf.h elfos.h gnu-user.h linux.h glibc-stdint.h ia64/sysv4.h ia64/linux.h"
-	tmake_file="${tmake_file} ia64/t-ia64 t-libunwind"
+	tmake_file="${tmake_file} ia64/t-ia64 ia64/t-glibc t-libunwind"
 	target_cpu_default="MASK_GNU_AS|MASK_GNU_LD"
 	;;
 ia64*-*-hpux*)
@@ -2057,6 +2060,7 @@
 		;;
 	    *)
 		tm_file="${tm_file} rs6000/linux.h glibc-stdint.h"
+		tmake_file="$tmake_file rs6000/t-linux"
 		;;
 	esac
 	case ${target} in
